### ISSUE TITLE
About the final XBNBlock-P2 module

### ISSUE THREAD

**Issue by yahuyang** [huangleiBuaa/XBNBlock#4]
![image](https://user-images.githubusercontent.com/73293451/168943358-4aa9337d-1482-4cac-ab78-a96ff9304b25.png)

Dear Professor Huang:
       Hello! Is the final XBNBlock-P2 module the following code?

    def GroupNorm(num_features, num_groups=32, eps=1e-5, affine=True, *args, **kwargs):
        if num_groups>num_features:
            print('------arrive maxum groub numbers of:', num_features)
            num_groups=num_features
        return nn.GroupNorm(num_groups, num_features, eps=eps, affine=affine)

    class Bottleneck_XBNBlock_P2(nn.Module):
            expansion = 4
        
            def __init__(self, inplanes, planes, stride=1, downsample=None):
                super(Bottleneck_XBNBlock_P2, self).__init__()
                self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
                self.bn1 = nn.BatchNorm2d(planes)
                self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
                self.bn2 = GroupNorm(planes)
                self.conv3 = nn.Conv2d(planes, planes * 4, kernel_size=1, bias=False)
                self.bn3 = nn.BatchNorm2d(planes * 4)
                self.relu = nn.ReLU(inplace=True)
                self.downsample = downsample
                self.stride = stride
    
            def forward(self, x):
                residual = x
        
                out = self.conv1(x)
                out = self.bn1(out)
                out = self.relu(out)
        
                out = self.conv2(out)
                out = self.bn2(out)
                out = self.relu(out)
        
                out = self.conv3(out)
                out = self.bn3(out)
        
                if self.downsample is not None:
                    residual = self.downsample(x)
        
                out += residual
                out = self.relu(out)
        
                return out

I read your code and only this class matches the module above.
Also, I use the standard Bottleneck in my work: conv 1*1->bn->relu->conv3*3->bn->relu. The structure of this Bottleneck is different from the one in your article. Can I replace the bn after 3*3 convolution with BFN(GN)?
Thank you very much.

**Comment by huangleiBuaa** [comment]
Hi, yes, the module you list is the XBNBlock-P2. You can replace the BN (after 3*3 convolution) with BFN in the standard 'basicBlock' (based on your description, you have only 2 conv, I am not sure whether you forget the other 1*1 conv), technicually. However, based on my experience on the Bottleneck, it seems that repalce the BN (after 1*1 convolution) with BFN should be better in your standard 'basicBlock'.

**Comment by yahuyang** [comment]
非常感谢黄老师的耐心解答。
我的Bottleneck如下图：
![image](https://user-images.githubusercontent.com/73293451/170440908-add22b64-b29b-4419-bddc-d052546b4442.png)
确实只有两个Conv。我前几天将3*3convolution之后的BN换成了BFN，效果确实没有以前好了。我再试试1*1convolution之后的BN换成BFN。
谢谢黄老师。

**Comment by Sakura0v01** [comment]
> 
同学你好，我想请问一下关于下载下来的代码ResNext_XBNBlock_P2.py中model_urls没有定义是如何解决的
![L0HKCT`5ZNM`O53AQ8XGXED](https://user-images.githubusercontent.com/112566422/224474891-4dbadbfe-54fe-4368-bacb-d05d3f2939d6.png)
